### DISCUSSION TITLE
When we can expect laravel/framework package to have return types build in?

### DISCUSSION THREAD

**Discussion by mikield** [laravel/framework#53089] [General]
It's 2024, and PHP 8.4 has been released.

Strong typed codebases have been proven stability over non strong typed code.

I am asking myself, when we can see such big tool as Laravel that's used by billions of web apps to be strong typed?!

Do Laravel team has some plans on it? 
If yes - I would like to contribute :)

**Comment by ghost** [comment]
Never:)

**Reply by ghost** [reply]
I saw ints in dock block and return as Builder.

**Reply by mikield** [reply]
That's very sad actually :(

**Reply by carlosmintfan** [reply]
Make a PR :)

**Reply by browner12** [reply]
I wouldn't waste your time with a PR. they are well aware of return types and have explicitly decided not to do them yet, but I'm sure it's something they're thinking about often.

If they would decide to do it, they'd probably handle it internally and do it in one fell swoop.

**Reply by mikield** [reply]
I think they will do it internally too, cause of big official package dependencies, and huge update cycle that would be hard to manage.

**Reply by ghost** [reply]
The flows will break. That is why this will not be done, except for putting mixed among other "known" result types. I tired it and like I said, instead of int I got Bulder. So best case scenario:

function x(...): mixed|int|Builder|...etc

**Reply by ghost** [reply]
Also when I did CR for their PR/MR the "king" rolled back the typed params and returns if the owner of the PR listened to me and committed return types and types params.

**Reply by mikield** [reply]
IDK even if we can fork laravel and make it strong typed, because this will not be laravel anymore :/

**Reply by ghost** [reply]
@mikield you will break it by putting strong typings

**Reply by mikield** [reply]
@marius-mcp if it will be a fork, I can anytime delete the code and nobody will even notice it xD
But if it will succeed - we can discuss merging strategy to Laravel core.

**Reply by ghost** [reply]
Good luck. It would be a titanic work and you will be forced to use alot of mixed return types and param types. Example forwardCallTo from model to eloquent builder that ends up in query builder.

**Reply by mikield** [reply]
Actually, I do not know if I will do that, that was just an idea, if I will hit some big wall - I will just throw that away, too much effort for 0 outcome :)

**Reply by ghost** [reply]
Welcome to the club. I had the same ideas as you.

**Reply by mikield** [reply]
@marius-mcp will take opportunity here to share with you my thoughts about Laravel and a nice (IMHO) solution to a problem that I do not like.
https://www.threads.net/@mikield/post/DA8W5V4IGyD

**Reply by carlosmintfan** [reply]
I just meant if they have declared the wrong return type in the docblock (int, but it returns builder) then you could make a PR, not the whole stuff replacing docblock by php return :)

**Reply by ghost** [reply]
@carlosmintfan 
Dockblock update with int|false|Builder is useless anyway.  A return type will not be accepted.
```php
\Illuminate\Database\Eloquent\Model::incrementOrDecrement
        return false;
        return tap($this->setKeysForSaveQuery($query)->{$method}($column, $amount, $extra), function () use ($column) {
            $this->syncChanges();

            $this->fireModelEvent('updated', false);

            $this->syncOriginalAttribute($column);
        });

     * @return \Illuminate\Database\Eloquent\Builder
     */
    protected function setKeysForSaveQuery($query)
    {
        $query->where($this->getKeyName(), '=', $this->getKeyForSaveQuery());

        return $query;
    }
```

**Reply by ghost** [reply]
@mikield I suggest you to not pick the bear. Leave laravel to rest in peace. If you don't like it do not use it. You will not be able to change anything. Others have tried and were bombarded by the community just like in a religious riot.

**Reply by mikield** [reply]
Do only me feels like we need here an RuntimeException?
Like, if you expect your code not to run - throw new Exception that will allow user to catch it and write "catch" case code.

![Знімок екрана 2024-10-10 о 18 25 48](https://github.com/user-attachments/assets/a85e9b11-f95c-4ce0-ba75-b17a92a6b5ef)

**Reply by mikield** [reply]
@marius-mcp I am coding only laravel for like 6+ years now, so I can not say that I do not like it, I want the best for the tool that I like :)

But I understand your point, Laravel is kinda Apple from PHP dev for some people :)

**Reply by ghost** [reply]
@mikield good thing is that the packages for laravel can be written ~ 90% with return types and typed params. Also the projects that we build using laravel/lumen.

**Reply by puklipo** [reply]
There is precedent for very painful breaking changes destroying communities.
The only solution is to introduce types little by little with each major version upgrade.
Newer packages that don't need to worry about breaking changes are well-typed.

**Reply by ghost** [reply]
@puklipo that is why I said never. At first I thought it is possible, but after more than 3 years working with it, I understood it is not possible because of the backward compatibility issues..

**Comment by macropay-solutions** [comment]
The reason it Returns Eloquent builder instead of int: https://github.com/laravel/framework/issues/56664

**Comment by macropay-solutions** [comment]
@mikield we started applying PSR12 but we did not touch the return types or the parameters types.

**Comment by macropay-solutions** [comment]
@mikield We saw some attempts on putting return types that generated bugs and were either patched or reverted in version 12. So the community is trying to improve the framework but the legacy is a big stopper for that.

**Reply by shaedrich** [reply]
That sadly sums it up pretty well 👍🏻

**Reply by macropay-solutions** [reply]
But that did not stop us from using PSR12 in our projects. Also when extending Laravel classes we could put the return type even if the parent did not had one.
Of course putting mixed does not help much but anyway...

**Reply by macropay-solutions** [reply]
The other half of the problem are the typed parameters on functions definition.

**Reply by shaedrich** [reply]
We might just have to tie these changes closer to [`/types`](https://github.com/laravel/framework/tree/12.x/types)

**Reply by macropay-solutions** [reply]
@shaedrich what are those? tests that test the return type? What can they bring as a plus value to the framework if you have an app that returns different type vs the framework?

**Reply by shaedrich** [reply]
I haven't worked with them yet so I don't know, but this can at least cover a good chunk.

**Reply by mbabker** [reply]
> the community is trying to improve the framework but the legacy is a big stopper for that

It's not even legacy.  The framework was designed as a duck-typed framework, and tries to provide functionality that isn't well supported in PHP as a whole through emulated method/function overloading (see all the stuff where return values change based on whether a param is given or stuff like the query builder where you can omit the `$operator` argument).  The end result is that when there is type information declared, it's unreliable at best and downright misleading at worst.  Two of the issues fixed with 12.26 patch releases show just that:

- https://github.com/laravel/framework/pull/56779 - The documented return type omitted the base guard interface
- https://github.com/laravel/framework/issues/56772 - The default implementation of the session contract violates the interface (which cascades down to helpers like `csrf_token()`) and can return an unexpected null value

No matter how much people push for it, Laravel cannot be made into a strongly typed framework because the framework is by design too dynamic to be fully typed.

**Reply by macropay-solutions** [reply]
@mbabker  https://github.com/laravel/framework/discussions/53089#discussioncomment-10894091
>Never:)

As a side note, if you duplicate all functions with multi-return you could make a framework that is fully typed...but it will be 50 times bigger in size.

As a first step, only functions that have single return type should be enforced.
Second step would be duplicating the functions that don't have typed params that are known for sure types, and deprecate the old ones.

```php

public function old($param1, $param2) {
    if (...) { //test params to be of the type required in new function
        return $this->new($param1, $param2);
    }

    .. old function body
}
```
This would be a backward compatible approach.

But still not all the framework can be made strong typed.

**Reply by Sophist-UK** [reply]
Yes - PHP doesn't support overloading natively, and thus Laravel's overloading and strict typing are not compatible. So you would need to have different named methods for each overloading variant - but presumably Rector could refactor any existing calls - but as just stated it would make the framework significantly bigger (perhaps 2x-5x bigger - not sure it would be 50x).

**Reply by mikield** [reply]
> @mbabker [#53089 (comment)](https://github.com/laravel/framework/discussions/53089#discussioncomment-10894091)
> 
> > Never:)
> 
> As a side note, if you duplicate all functions with multi-return you could make a framework that is fully typed...but it will be 50 times bigger in size.
> 
> As a first step, only functions that have single return type should be enforced. Second step would be duplicating the functions that don't have typed params that are known for sure types, and deprecate the old ones.
> 
> ```
> public function old($param1, $param2) {
>     if (...) { //test params to be of the type required in new function
>         return $this->new($param1, $param2);
>     }
> 
>     .. old function body
> }
> ```
> 
> This would be a backward compatible approach.

Or we can start building ecosystem around fresh and more modern frameworks, like [tempest](https://tempestphp.com/)

**Reply by macropay-solutions** [reply]
@mikield we started with a different approach from laravel 10 already. 
- No new features in core (all new features should come via external packages)
- PSR12 as much as possible
- Fix all known bugs (security issues like the [eager load issue](https://github.com/laravel/framework/issues/51825) that was NOT fixed in laravel) in a backward compatible manner
- Resolve all macroable classes from container instead of using new (to allow overriding as macros don't support that)
- Resolvable classes also with list array as parameters
- Needed changes for PHP 8.4 while keeping 8.1 compatibility.
- etc.

Update
tempest looks like a new symfony with routes registered via Attributes (scattered in controllers) (analog validators) and
seems like a marketing rule: "Don't make it better, make it different!"

**Reply by mbabker** [reply]
> tempest looks like a new symfony with routes registered via Attributes (scattered in controllers)

You know, not everything that supports attributes for configuration is Symfony.  Otherwise, Laravel itself with all of its newer attributes might as well be Symfony as well (not to mention things like Filament which also support them).

**Reply by mikield** [reply]
Tempest takes best of Laravel and Symfony.
Auto discovery out of the box.
No onion architecture.
Hardly based on discoveries.
Objects and strict types.

Not promoting it, but at first when I saw it - I thought, finaly someone made a framework that is minimal from the init design, and extendable with simple feature as discovery.

Nowdays every PHP framework uses Reflection and it's ok for new PHP :)

**Reply by macropay-solutions** [reply]
@mikield 
Thank you for the link. Interesting... We will see if we can port our out of the box rest/crud lib suite into tempest, but at first glance, if in Eloquent it took us 3 years to have a MVP, in tempest that seems impossible because Eloquent features would need porting as well.

Frameworks should shorten the delivery time of features (fully working without bugs, security issue or uncovered corner cases) for end clients. If they manage to do that, than it is ok no matter if they follow industry theories or not. 

Increasing billable hours for end clients and pleasing developers should be avoided. Clients pay so, their needs should drive the development. A code that can't adapt / change according to the client's needs does not have any value.

Lately Go is in fashion, it is faster ...until it waits for a slow DB query. Then it is as slow as any language...

**Comment by Sophist-UK** [comment]
As is very common in the real world (rather than theory) there is a balance to be had here. In general terms, introducing strict types are going to create (at least ***potentially***) breaking changes - and there are three separate situations to consider:

* Strict parameter types
* Strict return types
* `declare(strict_types=1);`

So you might introduce these at separate times and in separate ways.

Breaking changes are painful - but sometimes the pain is justified in order to provide a future path. Compare PHP4 with PHP8 - and think about how many breaking changes were successfully introduced to get from one to the other - and they were done because they were necessary for the long-term success of PHP. Breaking changes necessary for the long-term success of Laravel are equally justified IMO, but of course which changes are "necessary" is going to be subjective.

There are many many ways to mitigate / reduce the pain: 
* You can deprecate a major version ahead and provide warnings (in non-production environments)
* You can use tools like Rector to automate refactoring for some changes in order to reduce the effort to migrate to a new major version. If Laravel were to provide Rector rulesets that might go a long long way to making such changes palatable.

So, for example, in Laravel 13 you could introduce static-analysis and/or or run-time and/or automated test code to test parameter types and warn users when their code will create a breaking change in L14, and you could provide a Rector ruleset and an artisan command to refactor such code.

And of course, some parts of Laravel are used in a greater number of different ways than others, so introducing strict typing in lower levels and functions which are normally internal only will create less pain than for areas which are used widely and in many different ways.

And finally, new calls are introduced in many releases, and although these are a very small proportion of the overall code base, they could (and perhaps should) be introduced with strict typing wherever possible.

**Reply by mikield** [reply]
Introducing breaking changes in major versions is a common practice.
We should see significant difference between Laravel 12 and Laravel 11.
That should not be just a "here is a pack of some new features, and that's a new major version".

The framework should evolve, be modern with the language it's build on top of.
Strict types are not something new for PHP. Not even for Web world, everybody now loves TypeScript instead of Javascript, for its strictness.

Relying on framework when building a product that should work and be stable for 10+ years - and getting no types in functions arguments.
Having `mixed` everywhere it can be.
Zero consistency in return types.
I can not even extend models query builder without pain.
And I want to extend query builder, I want to be able to write clean code, instead of "magic code", that is handled by some acception.
That's why, when I query a model I do `Model::query()->get()` instead of `Model::get()`, because it explicitly tells that I am running a query, that gives me autocomplete without need of a separate plugin for my IDE.

And in the end it is very sad that nothing will change, because it gives 0 income for products of Laravel team, and gives a big backwards incompatibility.

**Reply by Sophist-UK** [reply]
Yes - when L12 was announced, one of my first thoughts was that zero breaking changes should not have been an objective - whilst if a release happens to have zero breaking changes that is a good thing, deliberately restricting changes to those that have zero breaking changes and avoiding significant improvements that would have breaking changes is NOT a good thing IMO.

**Reply by macropay-solutions** [reply]
@Sophist-UK the huge changes from 11 needed 2 major versions for bug fixing. That was the reason.

**Comment by GavG** [comment]
I expect we'd need Generics in PHP first, which may be a few years away still.

Laravel has a lot of instances where a functions return type isn't known until runtime.

Not least by the very nature of IoC containers, but also with things like callback evaluations, the Conditionable trait for example.

While it would be possible to type certain parts of the framework, it's likely assumed that this would be futile as the chain of typing is quickly broken where many parts of the framework are used together.

Still, I hope to see it one day🤞

**Reply by Sophist-UK** [reply]
I would say that any typing is better than none - so let's do it where it make sense and not where it doesn't. (But then again I may be a naive idiot.)
